### PR TITLE
[ReadHandler] ReportScheduler Injection

### LINKED ISSUE #28183: [ReadHandler] Allow custom ReportScheduler Injection
Currently, we are choosing between unsynchronized and synchronized report scheduler by a define in CHIPConfig.h :
`CHIP_CONFIG_SYNCHRONOUS_REPORTS_ENABLED`

We could allow users to define their own scheduler by allowing injection from the app layer by passing a pointer to a reportscheduler implementaiton in the Server::Init()

### PR DESCRIPTION
**Pull request by lpbeliveau-silabs**
Allowed injection of ReportScheduler from app layer without having a static default scheduler in place in the Server instance.

The default scheduler needs to be manually is still instantiated in CommonCaseDeviceServerInitParams. It can also be injected in the the server's initParams like such:

```
// Create initParams with SDK example defaults here
static chip::CommonCaseDeviceServerInitParams initParams;
// Report scheduler and timer delegate instance
static chip::app::DefaultTimerDelegate sTimerDelegate;
static chip::app::reporting::ReportSchedulerImpl sReportScheduler(&sTimerDelegate);
initParams.reportScheduler = &sReportScheduler;
```

This resolves the issue: https://github.com/project-chip/connectedhomeip/issues/28183

### CODE REVIEW

File: src/app/server/Server.h [hunk 1]
```diff
@@ -594,12 +596,7 @@ class Server
     app::DefaultAttributePersistenceProvider mAttributePersister;
     GroupDataProviderListener mListener;
     ServerFabricDelegate mFabricDelegate;
-    app::DefaultTimerDelegate mTimerDelegate;
-#if CHIP_CONFIG_SYNCHRONOUS_REPORTS_ENABLED
-    app::reporting::SynchronizedReportSchedulerImpl mReportScheduler;
-#else
-    app::reporting::ReportSchedulerImpl mReportScheduler;
-#endif
+    app::reporting::ReportScheduler * mReportScheduler;
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/server/Server.h:599]: > However, won't that lead to allocating two Schedulers in the event of the injection of a custom one?

If CommonCaseDeviceServerInitParams is not used, then no....

I think the setup should be:

1. The actual server code always expects a pointer in init params.  It does not null-check during init.
2. CommonCaseDeviceServerInitParams provides a pointer to one of its things by default.

Then if you use CommonCaseDeviceServerInitParams you can still customize but then do get both implementations.  If you don't use CommonCaseDeviceServerInitParams (as the docs say you should not do as a non-example application), then it should all get optimized out, I would think.

[the remark this answers is not in the document]
Comment (lpbeliveau-silabs) [reply src/app/server/Server.h:611]: It seems adding the static scheduler did lead to a significant increase in examples. 

While we do recommend not to use CommonCaseDeviceServerInitParams, their are still customers who base their products on the sample apps so they would be impacted, not to mention our tests are currently failing due to the size increase.

I can remove the instances in app and hope that makes enough of course.

File: src/app/server/Server.h [hunk 2]
```diff
@@ -146,6 +142,10 @@ struct ServerInitParams
     // Operational certificate store with access to the operational certs in persisted storage:
     // must not be null at timne of Server::Init().
     Credentials::OperationalCertificateStore * opCertStore = nullptr;
+    // Required, if not provided, the Server::Init() WILL fail. The default reportScheduler implementation will be used when using
+    // CommonCaseDeviceServerInitParams, but the injection of a custom scheduler requires using the ServerInitParams base class to
```
Comment (bzbarsky-apple) [inline src/app/server/Server.h:146]: You shouldn't even mention CommonCaseDeviceServerInitParams here.  People should not be using it.

File: src/lib/core/CHIPConfig.h [hunk 3]
```diff
@@ -1573,7 +1573,7 @@ extern const char CHIP_NON_PRODUCTION_MARKER[];
  * various subscribers.
  */
 #ifndef CHIP_CONFIG_SYNCHRONOUS_REPORTS_ENABLED
-#define CHIP_CONFIG_SYNCHRONOUS_REPORTS_ENABLED 0
+#define CHIP_CONFIG_SYNCHRONOUS_REPORTS_ENABLED 1
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPConfig.h:1576]: This needs to stay 0 for now, no?  Why is this being changed in this PR?

**Review by bzbarsky-apple — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Pull request merged**